File: docs/experiments/grok-app/native-v2/android/app/src/main/java/tv/whisco/app/LocalStore.kt
package tv.whisco.app

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.datastore.preferences.core.edit
import androidx.datastore.preferences.core.stringPreferencesKey
import androidx.datastore.preferences.preferencesDataStore
import kotlinx.coroutines.flow.first
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.util.UUID

private val Context.dataStore by preferencesDataStore("whisco")
private val KEY = stringPreferencesKey("state")
private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }

class LocalStore {
    var ready by mutableStateOf(false)
    var onboarded by mutableStateOf(false)
    var face by mutableStateOf(FaceDto())
    var faces by mutableStateOf(listOf(FaceDto()))
    var country by mutableStateOf("BH")
    var favoriteChannels by mutableStateOf(listOf<ChannelDto>())
    var watchlist by mutableStateOf(listOf<TitleDto>())
    var resume by mutableStateOf(listOf<ResumeDto>())
    var recentSearches by mutableStateOf(listOf<String>())
    var soundEnabled by mutableStateOf(false)

    val favoriteIds: Set<String> get() = favoriteChannels.map { it.id }.toSet()

    suspend fun load(context: Context) {
        val raw = context.dataStore.data.first()[KEY]
        if (raw != null) {
            runCatching {
                val s = json.decodeFromString<Persisted>(raw)
                onboarded = s.onboarded
                face = s.face
                faces = s.faces.ifEmpty { listOf(s.face) }
                country = s.country
                favoriteChannels = s.favoriteChannels.ifEmpty {
                    s.favoriteIds.map { ChannelDto(id = it, name = it) }
                }
                watchlist = s.watchlist
                resume = s.resume
                recentSearches = s.recentSearches
                soundEnabled = s.soundEnabled
            }
        }
        ready = true
    }

    suspend fun persist(context: Context) {
        context.dataStore.edit {
            it[KEY] = json.encodeToString(snapshot())
        }
    }

    fun completeOnboarding(
        name: String,
        languages: List<String>,
        country: String,
        clockSource: String,
        kidsMode: Boolean,
    ) {
        face = FaceDto(
            id = "you",
            name = name.ifBlank { "You" },
            languages = languages.ifEmpty { listOf("Arabic", "English") },
            hue = 18f,
            clockSource = clockSource,
            kidsMode = kidsMode,
        )
        faces = listOf(face)
        this.country = country
        onboarded = true
    }

    fun setActiveFace(id: String) {
        faces.firstOrNull { it.id == id }?.let { face = it }
    }

    fun addFace(name: String, languages: List<String>, kidsMode: Boolean) {
        if (faces.size >= 4) return
        val next = FaceDto(
            id = UUID.randomUUID().toString(),
            name = name,
            languages = languages,
            hue = (10..340).random().toFloat(),
            clockSource = if (languages.contains("Arabic") && languages.contains("English")) "gulf" else "origin",
            kidsMode = kidsMode,
        )
        faces = faces + next
        face = next
    }

    fun updateActive(patch: FaceDto.() -> FaceDto) {
        face = face.patch()
        faces = faces.map { if (it.id == face.id) face else it }
    }

    fun removeFace(id: String) {
        if (faces.size <= 1 || id == "you") return
        faces = faces.filterNot { it.id == id }
        if (face.id == id) face = faces.first()
    }

    fun toggleFavorite(ch: ChannelDto) {
        favoriteChannels = if (favoriteChannels.any { it.id == ch.id }) {
            favoriteChannels.filterNot { it.id == ch.id }
        } else {
            (listOf(ch) + favoriteChannels).take(16)
        }
    }

    fun toggleSaved(t: TitleDto) {
        watchlist = if (watchlist.any { it.id == t.id || it.slug == t.slug }) {
            watchlist.filterNot { it.id == t.id || it.slug == t.slug }
        } else {
            (listOf(t) + watchlist).take(80)
        }
    }

    fun isSaved(id: String, slug: String = id) =
        watchlist.any { it.id == id || it.slug == slug }

    fun upsertResume(entry: ResumeDto) {
        resume = (listOf(entry) + resume.filterNot { it.id == entry.id }).take(24)
    }

    fun pushSearch(q: String) {
        val t = q.trim()
        if (t.length < 2) return
        recentSearches = (listOf(t) + recentSearches.filterNot { it.equals(t, true) }).take(8)
    }

    fun reset() {
        onboarded = false
        face = FaceDto()
        faces = listOf(FaceDto())
        country = "BH"
        favoriteChannels = emptyList()
        watchlist = emptyList()
        resume = emptyList()
        recentSearches = emptyList()
        soundEnabled = false
    }

    private fun snapshot() = Persisted(
        onboarded, face, faces, country,
        favoriteChannels.map { it.id }, favoriteChannels,
        watchlist, resume, recentSearches, soundEnabled,
    )

    @Serializable
    private data class Persisted(
        val onboarded: Boolean = false,
        val face: FaceDto = FaceDto(),
        val faces: List<FaceDto> = emptyList(),
        val country: String = "BH",
        val favoriteIds: List<String> = emptyList(),
        val favoriteChannels: List<ChannelDto> = emptyList(),
        val watchlist: List<TitleDto> = emptyList(),
        val resume: List<ResumeDto> = emptyList(),
        val recentSearches: List<String> = emptyList(),
        val soundEnabled: Boolean = false,
    )
}

File: docs/experiments/grok-app/native-v2/ios/WhiscoTV/WhiscoTV/LocalStore.swift
import Foundation
import Observation

@Observable
final class LocalStore {
    var onboarded: Bool
    var faces: [Face]
    var activeFaceId: String
    var country: GccCountry
    var watchlist: [CatalogTitle]
    var favoriteChannels: [Channel]
    var resume: [ResumeEntry]
    var recentSearches: [String]
    var soundEnabled: Bool

    var activeFace: Face {
        faces.first(where: { $0.id == activeFaceId }) ?? faces[0]
    }

    init() {
        let d = UserDefaults.standard
        onboarded = d.bool(forKey: "onboarded")
        country = GccCountry(rawValue: d.string(forKey: "country") ?? "BH") ?? .BH
        soundEnabled = d.bool(forKey: "sound")
        if let data = d.data(forKey: "faces"),
           let decoded = try? JSONDecoder().decode([Face].self, from: data),
           !decoded.isEmpty {
            faces = decoded
        } else {
            faces = [Face(id: "you", name: "You", languages: ["Arabic", "English"], hue: 18, clockSource: "gulf", kidsMode: false)]
        }
        activeFaceId = d.string(forKey: "activeFace") ?? "you"
        watchlist = Self.decode(d.data(forKey: "watchlist"), as: [CatalogTitle].self) ?? []
        favoriteChannels = Self.decode(d.data(forKey: "favCh"), as: [Channel].self) ?? []
        resume = Self.decode(d.data(forKey: "resume"), as: [ResumeEntry].self) ?? []
        recentSearches = Self.decode(d.data(forKey: "recents"), as: [String].self) ?? []
    }

    func persist() {
        let d = UserDefaults.standard
        d.set(onboarded, forKey: "onboarded")
        d.set(country.rawValue, forKey: "country")
        d.set(soundEnabled, forKey: "sound")
        d.set(activeFaceId, forKey: "activeFace")
        d.set(try? JSONEncoder().encode(faces), forKey: "faces")
        d.set(try? JSONEncoder().encode(watchlist), forKey: "watchlist")
        d.set(try? JSONEncoder().encode(favoriteChannels), forKey: "favCh")
        d.set(try? JSONEncoder().encode(resume), forKey: "resume")
        d.set(try? JSONEncoder().encode(recentSearches), forKey: "recents")
    }

    func completeOnboarding(name: String, languages: [String], country: GccCountry, clockSource: String = "origin", kidsMode: Bool = false) {
        faces = [Face(id: "you", name: name.isEmpty ? "You" : name, languages: languages.isEmpty ? ["Arabic", "English"] : languages, hue: 18, clockSource: clockSource, kidsMode: kidsMode)]
        activeFaceId = "you"
        self.country = country
        onboarded = true
        persist()
    }

    func setActiveFace(_ id: String) {
        activeFaceId = id
        persist()
    }

    func addFace(name: String, languages: [String], kidsMode: Bool) {
        guard faces.count < 4 else { return }
        let face = Face(id: UUID().uuidString, name: name, languages: languages, hue: Double.random(in: 10...340), clockSource: languages.contains("Arabic") && languages.contains("English") ? "gulf" : "origin", kidsMode: kidsMode)
        faces.append(face)
        activeFaceId = face.id
        persist()
    }

    func updateActive(_ patch: (inout Face) -> Void) {
        guard let i = faces.firstIndex(where: { $0.id == activeFaceId }) else { return }
        patch(&faces[i])
        persist()
    }

    func removeFace(_ id: String) {
        guard faces.count > 1, id != "you" else { return }
        faces.removeAll { $0.id == id }
        if activeFaceId == id { activeFaceId = faces[0].id }
        persist()
    }

    func toggleSaved(_ title: CatalogTitle) {
        if let i = watchlist.firstIndex(where: { $0.id == title.id }) {
            watchlist.remove(at: i)
        } else {
            watchlist.insert(title, at: 0)
            if watchlist.count > 80 { watchlist = Array(watchlist.prefix(80)) }
        }
        persist()
    }

    func isSaved(_ id: String) -> Bool { watchlist.contains(where: { $0.id == id }) }

    func toggleFavorite(_ channel: Channel) {
        if let i = favoriteChannels.firstIndex(where: { $0.id == channel.id }) {
            favoriteChannels.remove(at: i)
        } else {
            favoriteChannels.insert(channel, at: 0)
            if favoriteChannels.count > 16 { favoriteChannels = Array(favoriteChannels.prefix(16)) }
        }
        persist()
    }

    func isFavorite(_ id: String) -> Bool { favoriteChannels.contains(where: { $0.id == id }) }

    func upsertResume(_ entry: ResumeEntry) {
        resume.removeAll { $0.id == entry.id }
        resume.insert(entry, at: 0)
        if resume.count > 24 { resume = Array(resume.prefix(24)) }
        persist()
    }

    func pushSearch(_ q: String) {
        let t = q.trimmingCharacters(in: .whitespacesAndNewlines)
        guard t.count >= 2 else { return }
        recentSearches.removeAll { $0.caseInsensitiveCompare(t) == .orderedSame }
        recentSearches.insert(t, at: 0)
        if recentSearches.count > 8 { recentSearches = Array(recentSearches.prefix(8)) }
        persist()
    }

    func reset() {
        onboarded = false
        faces = [Face(id: "you", name: "You", languages: ["Arabic", "English"], hue: 18, clockSource: "gulf", kidsMode: false)]
        activeFaceId = "you"
        country = .BH
        watchlist = []
        favoriteChannels = []
        resume = []
        recentSearches = []
        soundEnabled = false
        persist()
    }

    private static func decode<T: Decodable>(_ data: Data?, as: T.Type) -> T? {
        guard let data else { return nil }
        return try? JSONDecoder().decode(T.self, from: data)
    }
}
